src/porcupine/checker.rs

```diff
@@ -0,0 +1,160 @@
+                    let (call, ret) = entry.lift();
+                    calls.push(CallEntry {
+                        call,
+                        ret,
+                        state: mem::replace(&mut state, new_state),
+                    });
+                    if let Some(front) = undecided.front_mut() {
+                        entry = front;
+                    } else {
+                        break;
+                    }
+                } else {
+                    // this state is visited before
+                    entry = entry.next_mut().unwrap();
+                }
+            } else {
+                // call entry has next
+                entry = entry.next_mut().unwrap();
+            }
+        } else {
+            // an undecided return found, meaning that a call considered done before this
+            // time point has to be revoked.
+            debug!("id={} return", entry.id);
+            if calls.is_empty() {
+                return CheckResult::Illegal;
+            }
+            let CallEntry {
+                mut call,
+                ret,
+                state: state0,
+            } = calls.pop().unwrap();
+            debug!("revoke call {}", call.id);
+            state = state0;
+            linearized.set(call.id as _, false);
+            entry = call.ref_mut();
+            call.unlift(ret);
+            // call entry has next
+            entry = entry.next_mut().unwrap();
+        }
+    }
+    CheckResult::Ok
+}
+
+/// Check history in parallel.
+///
+/// For each sub-history, spawn a task to test its linearizability.
+async fn check_parallel<M: Model>(
+    histories: Vec<Vec<Entry<M>>>,
+    verbose: bool,
+    killed: Arc<AtomicBool>,
+) -> (CheckResult, LinearizationInfo) {
+    let mut futures: FuturesUnordered<_> = histories
+        .into_iter()
+        .map(|subhistory| {
+            let killed = killed.clone();
+            async move { check_single::<M>(subhistory, verbose, killed) }
+        })
+        .collect();
+    let mut check_result = CheckResult::Ok;
+    while let Some(res) = futures.next().await {
+        if res as u8 > check_result as u8 {
+            check_result = res;
+            if matches!(check_result, CheckResult::Illegal) && !verbose {
+                killed.store(true, Ordering::Relaxed);
+                break; // collect linearizable prefix under verbose mode
+            }
+        }
+    }
+    (check_result, LinearizationInfo {})
+}
+
+pub(super) async fn check_operations<M: Model>(
+    history: Vec<Operation<M>>,
+    verbose: bool,
+    timeout: Option<Duration>,
+) -> (CheckResult, LinearizationInfo) {
+    let histories = <M as Model>::partition(history);
+    let killed = Arc::new(AtomicBool::new(false));
+    if let Some(duration) = timeout {
+        let killed1 = killed.clone();
+        thread::spawn(move || {
+            thread::sleep(duration);
+            killed1.store(true, Ordering::Relaxed);
+        });
```

Comment:
It is derived from the Go test to avoid the linearizability check taking too much time. Seems an async timer is not suitable since this check is blocking and does not involve any async operation. (How about removing the timeout part? I have never encountered a situation where the check lasted more than 1 second...?)

Reply:
I would also prefer to remove the timeout.